File: missingNumber.java
//solution1: hashset
public int missing(int[] array) {
	int n = arr.length+1;
	HashSet<Integer> set = new HashSet<>();
	for(int number : array)
		set.add(number);
	//actually = is not necessary
	for(int i = 1; i <= n; i++){
		if(!set.contains(i))
			return i;
	}
	return n;
}
//solution2: sum  a math trick
//1+2+..n = n*(n+1)/2
public int missing(int[] array) {
	int n = array.length + 1;
	//0L: 0 in long type
	long targetSum = (n+0L)*(n+1)/2;
	long actualSum = 0L;
	for(int number : array){
		actualSum += num;
	}
	return (int)(targetSum - actualSum);
}

//solution3: bit manipulaton

File: hashtable/missingNumber.py
# number from 0 - n, no duplicate, find the missing one
# solution1:
def missingNumber(nums):
	s = set(nums)
	for each in range(len(nums) + 1):
		if each not in s:
			return each
	return -1


# solution2:
def missingNumber(nums):
	sum = 0
	n = len(nums)
	for each in nums:
		sum += each

	return n * (n+1)//2 - sum




# number from 1 - n, no duplicate, find the missing one
def missingNumber(nums):
	s = set(nums)
	for each in range(1, len(nums) + 1):
		if each not in s:
			return each
	return -1
